src/gps/dateTime.c

```diff
@@ -133,3 +133,118 @@ tiny_millis_t getUptime() {
 int getUptimeAsInt() {
    return (int) getUptime();
 }
+
+void getDateTimeFromEpochMillis(DateTime *dateTime, millis_t millis) {
+
+   uint64_t SecondsSinceEpoch = millis / 1000;
+
+   uint64_t sec;
+   uint32_t quadricentennials, centennials, quadrennials, annuals/*1-ennial?*/;
+   uint32_t year, leap;
+   uint32_t yday, hour, min;
+   uint32_t month, mday, wday;
+   static const uint32_t daysSinceJan1st[2][13]=
+   {
+     {0,31,59,90,120,151,181,212,243,273,304,334,365}, // 365 days, non-leap
+     {0,31,60,91,121,152,182,213,244,274,305,335,366}  // 366 days, leap
+   };
+   /*
+     400 years:
+
+     1st hundred, starting immediately after a leap year that's a multiple of 400:
+     n n n l  \
+     n n n l   } 24 times
+     ...      /
+     n n n l /
+     n n n n
+
+     2nd hundred:
+     n n n l  \
+     n n n l   } 24 times
+     ...      /
+     n n n l /
+     n n n n
+
+     3rd hundred:
+     n n n l  \
+     n n n l   } 24 times
+     ...      /
+     n n n l /
+     n n n n
+
+     4th hundred:
+     n n n l  \
+     n n n l   } 24 times
+     ...      /
+     n n n l /
+     n n n L <- 97'th leap year every 400 years
+   */
+
+   // Re-bias from 1970 to 1601:
+   // 1970 - 1601 = 369 = 3*100 + 17*4 + 1 years (incl. 89 leap days) =
+   // (3*100*(365+24/100) + 17*4*(365+1/4) + 1*365)*24*3600 seconds
+   sec = SecondsSinceEpoch + 11644473600;
+
+   wday = (uint32_t)((sec / 86400 + 1) % 7); // day of week
+
+   // Remove multiples of 400 years (incl. 97 leap days)
+   quadricentennials = (uint32_t)(sec / 12622780800ULL); // 400*365.2425*24*3600
+   sec %= 12622780800ULL;
+
+   // Remove multiples of 100 years (incl. 24 leap days), can't be more than 3
+   // (because multiples of 4*100=400 years (incl. leap days) have been removed)
+   centennials = (uint32_t)(sec / 3155673600ULL); // 100*(365+24/100)*24*3600
+   if (centennials > 3)
+   {
+     centennials = 3;
+   }
+   sec -= centennials * 3155673600ULL;
+
+   // Remove multiples of 4 years (incl. 1 leap day), can't be more than 24
+   // (because multiples of 25*4=100 years (incl. leap days) have been removed)
+   quadrennials = (uint32_t)(sec / 126230400); // 4*(365+1/4)*24*3600
+   if (quadrennials > 24)
+   {
+     quadrennials = 24;
+   }
+   sec -= quadrennials * 126230400ULL;
+
+   // Remove multiples of years (incl. 0 leap days), can't be more than 3
+   // (because multiples of 4 years (incl. leap days) have been removed)
+   annuals = (uint32_t)(sec / 31536000); // 365*24*3600
+   if (annuals > 3)
+   {
+     annuals = 3;
+   }
+   sec -= annuals * 31536000ULL;
+
+   // Calculate the year and find out if it's leap
+   year = 1601 + quadricentennials * 400 + centennials * 100 + quadrennials * 4 + annuals;
+   leap = !(year % 4) && (year % 100 || !(year % 400));
```

Comment:
There is a method you can use to calculate this as opposed to duplicating the logic.  See `isLeapYear`